### PR TITLE
✨ Optimizations against client-side throttling

### LINKED ISSUE #1916: bug:  Controller Manager fails to install on OpenShift - Times out waiting for CRD install
### Describe the bug

I tried to install Kubestellar (plus Kubeflex) on OpenShift using 0.21.0 release. The Controller Manager install fails with 

I0319 14:59:40.253321       1 crds.go:76] "applying crd" logger="Binding" name="bindings.control.kubestellar.io"
E0319 14:59:40.268791       1 main.go:147] "error installing the CRDs" err="context deadline exceeded" logger="setup"

This install uses WDS of type HOST

I use this command to install the Controller Manager:

helm  upgrade --install -n wds2-system kubestellar oci://ghcr.io/kubestellar/kubestellar/controller-manager-chart --version 0.21.0 --set ControlPlaneName=wds2 --set APIGroups=workload.codeflare.dev

The full log is here:
[cm.log](https://github.com/kubestellar/kubestellar/files/14652779/cm.log)

### Steps To Reproduce

kflex init

kubectl apply -f https://raw.githubusercontent.com/kubestellar/kubestellar/v0.21.0/config/postcreate-hooks/kubestellar.yaml
kubectl apply -f https://raw.githubusercontent.com/kubestellar/kubestellar/v0.21.0/config/postcreate-hooks/ocm.yaml

kflex create imbs1 --type vcluster -p ocm

helm --kube-context imbs1 upgrade --install status-addon -n open-cluster-management oci://ghcr.io/kubestellar/ocm-status-addon-chart --version v0.2.0-rc7

kflex create wds1 -p kubestellar

clusteradm --context imbs1 get token

On the WEC I run:
clusteradm join --hub-token <TOKEN> --hub-apiserver https://<hub endpoint> --cluster-name wec3

On the HUB I run:
clusteradm --context imbs1 accept --clusters wec3

kubectl config rename-context default/c109-e-eu-gb-containers-cloud-ibm-com:31392/system:serviceaccount:default:my-service-account wec3

// Create ClusterBinding
kubectl --context default/c115-e-us-south-containers-cloud-ibm-com:30829/system:serviceaccount:default:my-service-account apply -f - <<
EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kubeflex-manager-cluster-admin-rolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: kubeflex-controller-manager
  namespace: kubeflex-system
EOF

kubectl label controlplane wds2 kflex.kubestellar.io/cptype=wds

kflex ctx

helm  upgrade --install -n wds2-system kubestellar oci://ghcr.io/kubestellar/kubestellar/controller-manager-chart --version 0.21.0 --set ControlPlaneName=wds2 --set APIGroups=workload.codeflare.dev

kubectl get deployments.apps -n wds2-system kubestellar-controller-manager

### Expected Behavior

Controller Manager installs and in Ready state.

### Additional Context

_No response_

### LINKED ISSUE #1953: bug: `GetClient()` is triggering client-side throttling
### Describe the bug

Looking at lines 25 and 33 of the attached log, we can see two groups of client-side throttling.
[throttling.log](https://github.com/kubestellar/kubestellar/files/14724952/throttling.log)

After some initial investigation, it turns out that they are triggered by
https://github.com/kubestellar/kubestellar/blob/8d8561a315489dfc79891288a94b437349f27fd3/cmd/controller-manager/main.go#L123
and
https://github.com/kubestellar/kubestellar/blob/8d8561a315489dfc79891288a94b437349f27fd3/cmd/controller-manager/main.go#L132
, then in turn trace back to
https://github.com/kubestellar/kubestellar/blob/8d8561a315489dfc79891288a94b437349f27fd3/pkg/client/client.go#L43
which returns a RESTMapper.

It looks like this RESTMapper is currently not used anywhere. Maybe we can just remove it?

### Steps To Reproduce

Run the binding controller against a cluster with hundreds of GVRs.

### Expected Behavior

No throttling should appear when we `GetWDSKubeconfig` or `GetIMBSKubeconfig`.

### Additional Context

The 3rd group of throttling which begins at line 43 of the attached log is _not_ related to this issue. That group of throttling is being tracked and addressed in #1916.

### PR DESCRIPTION
**Pull request by waltforme**
## Summary
This PR introduces a few optimizations to address the client-side throttling.

From the binding controller's [log](https://github.com/kubestellar/kubestellar/files/14652779/cm.log), I identified two sources which independently trigger client-side throttling.
1. The API discovery of the WDS apiserver when the binding controller ensures the two KS CRDs (bindingpolicise, bindings). This source is tracked by #1916. This kind of client-side throttling is so severe that the binding controller just failed to start because of context timeout (for applying CRDs).  
1. The API discovery of the WDS apiserver when the binding controller retrieves the kubeconfigs of the WDS and the IMBS. This source is tracked by #1953.

This PR handles both sources, as follows.
1. For source 1, this PR replaces the full API discovery (call `ServerPreferredResources()`) with bare minimum discovery within a GroupVersion. Also, just for robustness, the context timeout for applying CRDs is extended.
1. For source 2, this PR kills the discoveries when the binding controller retrieves the kubeconfigs.

_In addition_ to the changes above, this PR also introduces another optimization to the extreme, which is reduce the full discovery to exactly once during the lifecycle of a binding controller.

All the changes of this PR are baked into quay.io/junatibm/controller-manager:1711133606. @cwiklik kindly helped me to test this image in his same environment as #1916. In the test, we were happy to see that the binding controller now starts successfully in seconds and all the throttling went away:
```
👉 Downloads $ oc cluster-info
Kubernetes control plane is running at https://c115-e.us-south.containers.cloud.ibm.com:30829

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
👉 Downloads $ oc -n wds2-system logs kubestellar-controller-manager-856c5ff888-w8rjn | head -n 100
I0323 16:34:38.681752       1 main.go:81] "Command line flag" logger="setup" name="add_dir_header" value="false"
I0323 16:34:38.681840       1 main.go:81] "Command line flag" logger="setup" name="alsologtostderr" value="false"
I0323 16:34:38.681849       1 main.go:81] "Command line flag" logger="setup" name="api-groups" value="workload.codeflare.dev,workloads.kueue.x-k8s.io,batch"
I0323 16:34:38.681857       1 main.go:81] "Command line flag" logger="setup" name="health-probe-bind-address" value=":8081"
I0323 16:34:38.681863       1 main.go:81] "Command line flag" logger="setup" name="kubeconfig" value=""
I0323 16:34:38.681869       1 main.go:81] "Command line flag" logger="setup" name="leader-elect" value="true"
I0323 16:34:38.681876       1 main.go:81] "Command line flag" logger="setup" name="log_backtrace_at" value=":0"
I0323 16:34:38.681882       1 main.go:81] "Command line flag" logger="setup" name="log_dir" value=""
I0323 16:34:38.681895       1 main.go:81] "Command line flag" logger="setup" name="log_file" value=""
I0323 16:34:38.681902       1 main.go:81] "Command line flag" logger="setup" name="log_file_max_size" value="1800"
I0323 16:34:38.681907       1 main.go:81] "Command line flag" logger="setup" name="logtostderr" value="true"
I0323 16:34:38.681913       1 main.go:81] "Command line flag" logger="setup" name="metrics-bind-address" value="127.0.0.1:8080"
I0323 16:34:38.681918       1 main.go:81] "Command line flag" logger="setup" name="one_output" value="false"
I0323 16:34:38.681923       1 main.go:81] "Command line flag" logger="setup" name="skip_headers" value="false"
I0323 16:34:38.681929       1 main.go:81] "Command line flag" logger="setup" name="skip_log_headers" value="false"
I0323 16:34:38.681935       1 main.go:81] "Command line flag" logger="setup" name="stderrthreshold" value="2"
I0323 16:34:38.681941       1 main.go:81] "Command line flag" logger="setup" name="v" value="2"
I0323 16:34:38.681948       1 main.go:81] "Command line flag" logger="setup" name="vmodule" value=""
I0323 16:34:38.681953       1 main.go:81] "Command line flag" logger="setup" name="wds-label" value=""
I0323 16:34:38.681959       1 main.go:81] "Command line flag" logger="setup" name="wds-name" value="wds2"
I0323 16:34:38.682581       1 listener.go:44] "Metrics server is starting to listen" logger="controller-runtime.metrics" addr="127.0.0.1:8080"
I0323 16:34:38.683009       1 main.go:122] "Getting config for WDS" logger="setup" name="wds2"
I0323 16:34:38.683030       1 kubeconfig.go:64] "using label" logger="setup" key="kflex.kubestellar.io/cptype" value="wds"
I0323 16:34:38.683438       1 kubeconfig.go:84] "waiting for cp with label" logger="setup" key="kflex.kubestellar.io/cptype" value="wds"
I0323 16:34:38.751916       1 main.go:128] "Got config for WDS" logger="setup" name="wds2"
I0323 16:34:38.751939       1 main.go:131] "Getting config for IMBS" logger="setup"
I0323 16:34:38.753961       1 kubeconfig.go:84] "waiting for cp with label" logger="setup" key="kflex.kubestellar.io/cptype" value="imbs"
I0323 16:34:38.812203       1 main.go:137] "Got config for IMBS" logger="setup" name="imbs1"
I0323 16:34:38.928888       1 controller.go:173] "Discovery" logger="Binding" numAPIResourceLists=149 err=null
I0323 16:34:38.928933       1 controller.go:141] "Parameters of the tuned client's token bucket rate limiter" logger="Binding" burst=300 qps=50
I0323 16:34:38.986814       1 crds.go:76] "applying crd" logger="Binding" name="bindingpolicies.control.kubestellar.io"
I0323 16:34:39.046510       1 crds.go:87] "crd established" logger="Binding" name="bindingpolicies.control.kubestellar.io"
I0323 16:34:39.052039       1 crds.go:76] "applying crd" logger="Binding" name="bindings.control.kubestellar.io"
I0323 16:34:39.108375       1 crds.go:87] "crd established" logger="Binding" name="bindings.control.kubestellar.io"
I0323 16:34:39.108414       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="v1"
I0323 16:34:39.108422       1 controller.go:283] "Ignoring APIResourceList" logger="Binding" groupVersion="apiregistration.k8s.io/v1"
I0323 16:34:39.108446       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="apps/v1"
I0323 16:34:39.108452       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="events.k8s.io/v1"
I0323 16:34:39.108460       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="authentication.k8s.io/v1"
I0323 16:34:39.108467       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="authorization.k8s.io/v1"
I0323 16:34:39.108474       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="autoscaling/v2"
I0323 16:34:39.108482       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="autoscaling/v1"
I0323 16:34:39.108492       1 controller.go:290] "Working on APIResourceList" logger="Binding" groupVersion="batch/v1" numResources=2
I0323 16:34:39.108800       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="certificates.k8s.io/v1"
I0323 16:34:39.108820       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="networking.k8s.io/v1"
I0323 16:34:39.108828       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="policy/v1"
I0323 16:34:39.108837       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="rbac.authorization.k8s.io/v1"
I0323 16:34:39.108844       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="storage.k8s.io/v1"
I0323 16:34:39.108852       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="storage.k8s.io/v1beta1"
I0323 16:34:39.108879       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="admissionregistration.k8s.io/v1"
I0323 16:34:39.108897       1 controller.go:290] "Working on APIResourceList" logger="Binding" groupVersion="apiextensions.k8s.io/v1" numResources=1
I0323 16:34:39.109329       1 controller.go:283] "Ignoring APIResourceList" logger="Binding" groupVersion="scheduling.k8s.io/v1"
I0323 16:34:39.109369       1 controller.go:283] "Ignoring APIResourceList" logger="Binding" groupVersion="coordination.k8s.io/v1"
I0323 16:34:39.109384       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="node.k8s.io/v1"
I0323 16:34:39.109393       1 controller.go:283] "Ignoring APIResourceList" logger="Binding" groupVersion="discovery.k8s.io/v1"
I0323 16:34:39.109405       1 controller.go:283] "Ignoring APIResourceList" logger="Binding" groupVersion="flowcontrol.apiserver.k8s.io/v1beta3"
I0323 16:34:39.109416       1 controller.go:283] "Ignoring APIResourceList" logger="Binding" groupVersion="flowcontrol.apiserver.k8s.io/v1beta2"
I0323 16:34:39.109450       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="proxy.open-cluster-management.io/v1alpha1"
I0323 16:34:39.109464       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="proxy.open-cluster-management.io/v1beta1"
I0323 16:34:39.109473       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="apps.openshift.io/v1"
I0323 16:34:39.109482       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="authorization.openshift.io/v1"
I0323 16:34:39.109491       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="build.openshift.io/v1"
I0323 16:34:39.109500       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="image.openshift.io/v1"
I0323 16:34:39.109508       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="oauth.openshift.io/v1"
I0323 16:34:39.109517       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="project.openshift.io/v1"
I0323 16:34:39.109578       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="quota.openshift.io/v1"
I0323 16:34:39.109588       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="route.openshift.io/v1"
I0323 16:34:39.109632       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="security.openshift.io/v1"
I0323 16:34:39.110027       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="template.openshift.io/v1"
I0323 16:34:39.110042       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="user.openshift.io/v1"
I0323 16:34:39.110051       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="packages.operators.coreos.com/v1"
I0323 16:34:39.110060       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="config.openshift.io/v1"
I0323 16:34:39.110069       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="operator.openshift.io/v1"
I0323 16:34:39.110077       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="operator.openshift.io/v1alpha1"
I0323 16:34:39.110085       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="acme.cert-manager.io/v1"
I0323 16:34:39.110093       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="admission.hive.openshift.io/v1"
I0323 16:34:39.110103       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="agent.open-cluster-management.io/v1"
I0323 16:34:39.110110       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="apiserver.openshift.io/v1"
I0323 16:34:39.110119       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="apps.open-cluster-management.io/v1"
I0323 16:34:39.110127       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="apps.open-cluster-management.io/v1beta1"
I0323 16:34:39.110135       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="apps.open-cluster-management.io/v1alpha1"
I0323 16:34:39.110143       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="cert-manager.io/v1"
I0323 16:34:39.110152       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="cloudcredential.openshift.io/v1"
I0323 16:34:39.110161       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="cluster.open-cluster-management.io/v1"
I0323 16:34:39.110169       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="cluster.open-cluster-management.io/v1beta2"
I0323 16:34:39.110177       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="cluster.open-cluster-management.io/v1beta1"
I0323 16:34:39.110185       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="cluster.open-cluster-management.io/v1alpha1"
I0323 16:34:39.110193       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="console.open-cluster-management.io/v1"
I0323 16:34:39.110201       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="console.openshift.io/v1"
I0323 16:34:39.110209       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="console.openshift.io/v1alpha1"
I0323 16:34:39.110217       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="crd.projectcalico.org/v1"
I0323 16:34:39.110225       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="dashboard.opendatahub.io/v1"
I0323 16:34:39.110234       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="dashboard.opendatahub.io/v1alpha"
I0323 16:34:39.110242       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="discovery.open-cluster-management.io/v1"
I0323 16:34:39.110251       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="discovery.open-cluster-management.io/v1alpha1"
I0323 16:34:39.110258       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="hive.openshift.io/v1"
I0323 16:34:39.110265       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="ibm.com/v1"
I0323 16:34:39.110274       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="ibm.com/v1alpha1"
I0323 16:34:39.110293       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="imageregistry.operator.openshift.io/v1"
I0323 16:34:39.110301       1 controller.go:287] "No need to watch per user input" logger="Binding" groupVersion="ingress.operator.openshift.io/v1"
```

## Related issue(s)
Fixes #1916 
Fixes #1953

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
Do any of the changes here reduce the functionality? Does any controller stop reacting to changes, or become slower in reacting to changes, in the set of discoverable resources?

**Comment by waltforme** [comment]
> Do any of the changes here reduce the functionality? Does any controller stop reacting to changes, or become slower in reacting to changes, in the set of discoverable resources?

No changes here reduce the functionality. Not stop or become slower, either.

File: pkg/binding/controller.go [hunk 1]
```diff
@@ -115,10 +121,19 @@ func NewController(parentLogger logr.Logger, wdsRestConfig *rest.Config, imbsRes
 		return nil, err
 	}
 
-	nGVRs, err := countGVRsViaAggregatedDiscovery(kubernetesClient)
+	// do the discovery, save the results, then discard the 'depleted' client
```
Comment (MikeSpreitzer) [inline pkg/binding/controller.go:124]: k8s.io/client-go maintains an HTTP/2 connection per _what_?

**Review by MikeSpreitzer — COMMENTED** [review]
Based only on visual inspection,
/lgtm
Thanks!

File: pkg/binding/controller.go [hunk 2]
```diff
@@ -179,9 +198,8 @@ func computeQPSFromNumGVRs(nGVRs int) float32 {
 		return rest.DefaultQPS
 	}
 	// in case too large, look at some value for reference
-	// https://github.com/kubernetes/kubernetes/pull/105520/files
-	if qps > 50.0 {
-		return 50.0
+	if qps > referenceQPSUpperBound {
```
Comment (ezrasilvera) [inline pkg/binding/controller.go:201]: Why ?  Don't we allow overriding with explicit larger number ?

File: pkg/crd/crds.go [hunk 3]
```diff
@@ -57,7 +57,7 @@ const (
 )
 
 func ApplyCRDs(ctx context.Context, dynamicClient dynamic.Interface, clientset kubernetes.Interface, clientsetExt apiextensionsclientset.Interface, logger logr.Logger) error {
-	ctxLimited, cancel := context.WithTimeout(ctx, 30*time.Second)
+	ctxLimited, cancel := context.WithTimeout(ctx, time.Minute)
```
Comment (ezrasilvera) [inline pkg/crd/crds.go:60]: Any special reason ?

File: pkg/binding/controller.go (hunk 2 above)
Comment (MikeSpreitzer) [reply pkg/binding/controller.go:201]: This PR is not changing the fact that this is not configured.
_I_ have not asked for making it configurable because I have not seen a need for adding that complexity.

File: pkg/binding/controller.go (hunk 1 above)
Comment (waltforme) [reply pkg/binding/controller.go:124]: Let me explain why a `disposableClient` is used here.

Before this PR, we pass binding controller's `kubernetesClient` to `func countGVRsViaAggregatedDiscovery` to do the discovery. After the discovery, the tokens could be 'depleted'. This depletion can be demonstrated by the attached cm.log file in the openning comment. If we look at lines 41 and lines 69 of the log, we can see that the `kubernetesClient` does two consecutive full (before this PR) discovery when applying KS CRDs. The first full discovery succeeded without showing throttling in the log (line 42). The second full discovery, however, triggers a group of throttling (starting with line 43), and leads to the context timeout, thus failure of start. The two full discoveries run the same section of code, but only the second discovery triggers throttling, that's because the first full discovery, perhaps plus part of the second full discovery, depleted `kubernetesClient`'s token bucket.

In this PR, `disposableClient` is assigned to the one-time full discovery. After the heavy-lifting job, `disposableClient` could be (based on # GVRs in WDS) depleted, but it is not returned by `NewController`, thus pop from the stack and disposed. By doing this, we make sure that binding controller's `kubernetesClient` is fresh with a full token bucket. This fresh `kubernetesClient` is now ready to do the two consecutive discoveries when applying CRDs. Also, the two consecutive discoveries are not full discovery anymore, they are just very light discoveries in one single GV.

File: pkg/crd/crds.go (hunk 3 above)
Comment (waltforme) [reply pkg/crd/crds.go:60]: This is a work around for the context timeout.

If we look at line 68 of the attached cm.log in the opening commit, we can see that the 30 seconds is used up. What @pdettori did as the first-aid is to extend this timeout. So even without changes in this PR, the controller can still (slowly) starts.

We can see that now the controller starts in a few seconds from the updated log (from 16:34:38.XXX to 16:34:39.XXX). But I still keep this work around just for robustness.

File: pkg/binding/controller.go (hunk 1 above)
Comment (MikeSpreitzer) [reply pkg/binding/controller.go:124]: I was reacting to the comment, questioning the implicit assumption about what the rate limiter is associated with; how do you know that the code in this PR causes two different rate limiters to be used? Certainly the HTTPS connection is shared (see https://github.com/kubernetes/client-go/blob/v0.28.2/transport/transport.go#L54).

I realize now that I mistakenly asked about the HTTPS connection when I _should_ have asked about the rate limiter. I have since looked, and found the answer: if you do not set the `RateLimiter` field of the `rest.Config` then you get a new rate limiter with each call to `kubernetes.NewForConfig` (regardless of what `*rest.Config` is passed --- N.B. that is a pointer type). You can see this at https://github.com/kubernetes/client-go/blob/v0.28.2/kubernetes/clientset.go#L473 and https://github.com/kubernetes/client-go/blob/v0.28.2/kubernetes/clientset.go#L481-L486 .

So, in short, I have answered my question about the comment. It is accurate.

File: pkg/binding/controller.go (hunk 2 above)
Comment (waltforme) [reply pkg/binding/controller.go:201]: I think the QPS should be upper-bounded somehow, so I looked around and found this reference value. Also, since we know that the `disposableClient` is going to do the full discovery, I just explicitly set the QPS to this upper bound for `disposableClient`.

Comment (ezrasilvera) [reply pkg/binding/controller.go:201]: I think it's a good idea to be able to override through external parameter. Even for just some special test cases where we want to remove the limit completely. 
But I don't think my comment should block this PR.  
Up to you @waltforme if you want to change this in future PR.

File: pkg/binding/controller.go (hunk 1 above)
Comment (waltforme) [reply pkg/binding/controller.go:124]: Thanks for the pointers! BTW, I am a bit confused about the word 'shallow' in https://github.com/kubernetes/client-go/blob/6030d798e44096d5ffb58e00802cadd81a74dcf0/kubernetes/clientset.go#L461. By creating a variable `configShallowCopy` after the dereference, don't we have a 'deep' copy of the content that `c` points to?
